### PR TITLE
Fix Issues #665 and #666: Parser and KKT domain mismatch fixes

### LINKED ISSUE #665: Parser truncates set element names containing hyphens in Table data
## Summary

The GAMS parser truncates set element names that contain hyphens when parsing Table data. For example, `20-bond-wt` becomes `20` in the parsed parameter data, but is correctly parsed as `20-` in the set declaration (though still missing `bond-wt`).

## Reproduction

**Model:** `data/gamslib/raw/ajax.gms`

**Original GAMS source:**
```gams
Set
   g 'paper grades'      / 20-bond-wt, 25-bond-wt, c-bond-ext, tissue-wrp /;

Table dempr(g,*) 'demand and prices'
                  demand  price
   20-bond-wt      30000     77
   25-bond-wt      20000     81
   c-bond-ext      12000     99
   tissue-wrp       8000    105;
```

**Parsed IR:**
```python
# Set g members - note truncation to '20-' instead of '20-bond-wt'
Set g members: ['20-', '25-', 'c-bond-ext', 'tissue-wrp']

# Parameter dempr values - note '20' instead of '20-' or '20-bond-wt'
dempr values:
  ('20', 'demand') = 30000.0
  ('20', 'price') = 77.0
  ('25', 'demand') = 20000.0
  ('25', 'price') = 81.0
  ("'c-bond-ext'", 'demand') = 12000.0
  ...
```

## Impact

1. **Set/Parameter mismatch**: Set `g` has element `'20-'` but parameter `dempr` references element `20` - these don't match
2. **Full name lost**: The complete element name `20-bond-wt` is lost entirely
3. **Quoted elements preserved**: Elements like `c-bond-ext` that get quoted are parsed correctly (as `'c-bond-ext'`)

## Root Cause

The parser appears to be treating `-` as a token separator in Table row labels, splitting `20-bond-wt` into separate tokens. The set declaration parsing handles this differently (preserving `20-`) but still truncates.

## Expected Behavior

- Set `g` should have members: `['20-bond-wt', '25-bond-wt', 'c-bond-ext', 'tissue-wrp']`
- Parameter `dempr` keys should be: `('20-bond-wt', 'demand')`, etc.

## Related

- Discovered during PR #664 review
- Similar issue may affect other models with hyphenated element names

### LINKED ISSUE #666: KKT generation: Domain mismatch when variable domain differs from equation domain
## Summary

The KKT generator produces stationarity equations with incorrect variable indices when a variable is declared over a subset but appears in equations indexed over a superset.

## Reproduction

**Model:** `data/gamslib/raw/chenery.gms`

**Original variable declarations:**
```gams
Variable
   e(i)    'quantity of exports'
   m(i)    'quantity of imports'
   g(t)    'foreign exchange cost of imports'   * t is subset of i
   h(t)    'foreign exchange value of exports'  * t is subset of i
```

**Key equation (trade balance):**
```gams
tb..  sum(t, g(t)*m(t) - h(t)*e(t)) =l= dbar;
```

**Generated stationarity equation (incorrect):**
```gams
stat_e(i).. ... + ((-1) * h(i)) * lam_tb + ... =E= 0;
stat_m(i).. ... + g(i) * lam_tb + ... =E= 0;
```

**Expected stationarity equation:**
```gams
stat_e(i).. ... + ((-1) * h(t))$t(i) * lam_tb + ... =E= 0;
stat_m(i).. ... + g(t)$t(i) * lam_tb + ... =E= 0;
```

Or alternatively using a sum:
```gams
stat_e(i).. ... + sum(t$sameas(t,i), (-1) * h(t)) * lam_tb + ... =E= 0;
```

## Problem Analysis

1. Variable `h` is declared over `t` (tradables subset)
2. Variable `e` is declared over `i` (all sectors)
3. In equation `tb`, `h(t)*e(t)` appears - both indexed by `t`
4. The stationarity equation for `e` is indexed by `i`
5. When computing ∂tb/∂e, the derivative of `h(t)*e(t)` w.r.t. `e(i)` should be:
   - `h(t)` when `i ∈ t` (i.e., when `t(i)` is true)
   - `0` otherwise
6. The current code emits `h(i)` which is invalid since `h` is not defined for all `i`

## Impact

- GAMS compilation error: domain mismatch
- The chenery model cannot be solved as MCP

## Root Cause Location

The issue is in the KKT stationarity equation generation code, likely in:
- `src/kkt/assemble.py` - where stationarity equations are constructed
- `src/kkt/stationarity.py` - if exists, stationarity-specific logic
- The code that maps constraint indices to variable indices during differentiation

## Fix Approach

When a term in a stationarity equation references a variable with a different (subset) domain:
1. Detect the domain mismatch between the equation index and the variable's declared domain
2. Either:
   a. Apply a conditional (`$t(i)`) to restrict the reference to valid domain elements
   b. Use `sum(t$sameas(t,i), ...)` to map indices correctly
   c. Keep the original variable index (`h(t)`) and wrap the term appropriately

## Related

- Discovered during PR #664 review
- Similar issues may affect other models with subset-indexed variables

### PR DESCRIPTION
**Pull request by jeffreyhorn**
## Summary

This PR fixes two issues:

### Issue #665: Parser truncates hyphenated elements

**Problem:** Identifiers starting with numbers and containing hyphens (e.g., `20-bond-wt`) were being truncated to just the leading number (`20`).

**Solution:** 
- Extended the preprocessor pattern to match number-starting hyphenated identifiers
- Added STRING token support in `dotted_label` grammar rule
- Extended `_token_text()` to strip quotes from escaped ID tokens
- Implemented range-based column matching algorithm for tables

**Files modified:**
- `src/ir/preprocessor.py`
- `src/ir/parser.py`
- `src/gams/gams_grammar.lark`
- `tests/unit/gams/test_parser.py`
- `tests/unit/test_special_identifiers.py`
- `tests/unit/test_multiple_alias_declaration.py`

### Issue #666: KKT domain mismatch with subset variables

**Problem:** When a variable is defined over a subset (e.g., `h(t)` where `t(i)` is a subset of `i`) and appears in a stationarity equation indexed over the superset (`stat_e(i)`), the code incorrectly produced `h(i)` instead of `h(t)`, causing GAMS Error 170.

**Solution:**
- Added `_preserve_subset_var_indices()` helper function to detect subset relationships
- Modified `_replace_indices_in_expr()` to preserve subset indices for variables

**Files modified:**
- `src/kkt/stationarity.py`
- `tests/integration/kkt/test_stationarity.py`

## Testing

All quality checks pass:
- `make typecheck` ✅
- `make lint` ✅
- `make format` ✅
- `make test` ✅ (3279 passed)

## Closes

- Closes #665
- Closes #666

### CODE REVIEW

File: tests/unit/test_special_identifiers.py [hunk 1]
```diff
@@ -199,26 +199,37 @@ def test_data_block_with_descriptions():
...
        light-ind  food+agr  heavy-ind
 food+agr      .1
 heavy-ind     .2        .1;"""
+    # Column headers should NOT be quoted; only row labels are quoted
     expected = """Table aio(i,i)
-       'light-ind'  'food+agr'  'heavy-ind'
+       light-ind  food+agr  heavy-ind
 'food+agr'      .1
 'heavy-ind'     .2        .1;"""
     assert normalize_special_identifiers(source) == expected
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/test_special_identifiers.py]: Fixed in 1bc100f. Added test_table_column_headers_with_plus_identifiers to verify the full parser pipeline correctly handles table column headers with + in identifiers like food+agr. The test confirms + is not misinterpreted as a continuation marker and all table data parses correctly.

File: tests/unit/gams/test_parser.py [hunk 2]
```diff
@@ -3311,6 +3311,69 @@ def test_hyphenated_multiline(self):
...
+        assert ("20-bond-wt", "machine-1") in model.params["prate"].values
+        assert ("25-bond-wt", "machine-2") in model.params["prate"].values
+        assert model.params["prate"].values[("20-bond-wt", "machine-1")] == 53
+        assert model.params["prate"].values[("c-bond-ext", "machine-2")] == 45
+
+    def test_table_column_headers_with_plus_identifiers(self):
+        """Test table with + in column headers doesn't misparse as continuation.
+
+        Issue #665: The + in identifiers like 'food+agr' must not be interpreted
+        as a table continuation marker. This verifies the full parser pipeline
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/gams/test_parser.py]: Fixed in e4793a1. Renamed the test to test_table_with_plus_in_row_labels and updated the docstring to clarify that this test covers + in row labels (which work correctly because the preprocessor quotes them), not column headers. Added a note documenting the limitation: column headers with + have limited support due to grammar constraints (the DESCRIPTION terminal doesn't match + and the table_continuation rule triggers).

File: tests/unit/test_special_identifiers.py [hunk 3]
```diff
@@ -199,26 +199,37 @@ def test_data_block_with_descriptions():
 
 
 def test_table_header_special_chars():
-    """Test quoting special characters in table column headers."""
+    """Test that table column headers are NOT quoted (Issue #665).
+
+    Column headers are kept unquoted because:
+    1. Quoting would make the first header look like a table description
+    2. The DESCRIPTION terminal matches multiple hyphenated words
+    The parser handles unquoted hyphenated headers via DESCRIPTION token parsing.
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/test_special_identifiers.py]: Fixed in e4793a1. Updated the docstring to clarify:
- Hyphenated headers (light-ind, heavy-ind) are parsed via the DESCRIPTION terminal
- The DESCRIPTION terminal does NOT match +, so headers containing + (like food+agr) have limited support and may trigger the table continuation rule
- Row labels with + are properly quoted and work correctly

This follows your suggested wording to distinguish hyphen-only headers from + headers.

File: src/ir/parser.py [hunk 4]
```diff
@@ -2002,19 +2028,29 @@ def _handle_table_block(self, node: Tree) -> None:
...
-                    dist = abs(token_col - col_pos)
-                    if dist < min_dist and dist <= 6:
-                        min_dist = dist
+                    # Determine the range this column owns
+                    # Start: this column's position (with small left tolerance)
+                    range_start = col_pos - 3  # Allow 3 chars left tolerance
+                    # End: next column's position (or infinity for last column)
+                    if idx + 1 < len(col_headers):
+                        next_col_pos = col_headers[idx + 1][1]
+                        range_end = next_col_pos - 1
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/ir/parser.py]: Fixed in 23c90f4. Replaced the overlapping range logic with non-overlapping ranges: each column now owns from just after the previous column's position (prev_col_pos + 1) to the next column's position (exclusive). The first column retains the small left tolerance (col_pos - 3). This eliminates the overlap that existed between `next_col_pos - 1` and `col_pos - 3` boundaries while maintaining correct column assignment for table values.

File: src/ir/parser.py [hunk 5]
```diff
@@ -2002,19 +2028,36 @@ def _handle_table_block(self, node: Tree) -> None:
...
+                    # End: next column's position (exclusive), or infinity for last
+                    if idx + 1 < len(col_headers):
+                        next_col_pos = col_headers[idx + 1][1]
+                        range_end = next_col_pos
+                    else:
+                        range_end = float("inf")
+                    # Check if value falls in this column's range
+                    if range_start <= token_col < range_end:
                         best_match = col_name
+                        break
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/ir/parser.py]: Fixed in f72af32. Added a sequential fallback when no range matches: if a value token falls outside all column ranges (e.g., left of the first header's col_pos - 3), it is now assigned to the next unused column instead of being silently dropped. This prevents data loss in tables where header lines are heavily indented relative to numeric values.

File: tests/unit/test_multiple_alias_declaration.py [hunk 6]
```diff
@@ -125,17 +125,21 @@ def test_multiple_alias_statements():
...
     """
     model = parse_model_text(source)
 
     assert "i" in model.sets
-    # Quoted identifiers are stored with quotes
-    assert "'i-alias'" in model.aliases
-    assert model.aliases["'i-alias'"].target == "i"
+    # Issue #665: Quotes are stripped from escaped identifiers
+    assert "i-alias" in model.aliases
+    assert model.aliases["i-alias"].target == "i"
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/test_multiple_alias_declaration.py]: Fixed in a839bbf. Added _quote_symbol() helper function and updated all symbol emission paths (aliases, sets, parameters, set assignments) to quote names containing special characters like '-' or '+'. Added test_alias_with_special_characters to verify that aliases like 'i-alias' are emitted as Alias(i, 'i-alias'); for valid GAMS output.

File: src/ir/parser.py [hunk 7]
```diff
@@ -372,14 +384,26 @@ def parse_model_file(path: str | Path) -> ModelIR:
...
-        return value[1:-1].strip()
+    if len(value) >= 2:
+        if token.type == "STRING":
+            return value[1:-1].strip()
+        # ID tokens can also be quoted via ESCAPED pattern (e.g., '20-bond-wt').
+        # For escaped identifiers, preserve inner whitespace as part of the name.
+        if token.type == "ID" and (
+            (value[0] == "'" and value[-1] == "'") or (value[0] == '"' and value[-1] == '"')
+        ):
+            return value[1:-1]
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply src/ir/parser.py]: Fixed in a839bbf. Rather than limiting quote-stripping in _token_text(), I've ensured all emitters consistently quote/sanitize symbol names that require escaping. Added _quote_symbol() helper and updated emit_original_sets(), emit_original_parameters(), emit_original_aliases(), emit_computed_parameter_assignments(), and emit_set_assignments() to quote any symbol name containing '-' or '+'. This approach keeps the parser logic simple (stripped identifiers) while ensuring valid GAMS output.

File: docs/issues/completed/ISSUE_665_parser-truncates-hyphenated-elements.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+### 1. Extended preprocessor pattern (`src/ir/preprocessor.py`)
+
+Added support for number-starting hyphenated identifiers:
+```python
+# Before: Only matched letter-starting identifiers
+pattern = r"\b([a-zA-Z_][a-zA-Z0-9_]*(?:[-+][a-zA-Z0-9_]+)+)\b"
+
+# After: Also matches number-starting identifiers like 20-bond-wt
+pattern = r"\b((?:[a-zA-Z_][a-zA-Z0-9_]*(?:[-+][a-zA-Z0-9_]+)+)|(?:[0-9]+[-][a-zA-Z0-9_]+(?:[-+][a-zA-Z0-9_]+)*))\b"
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply docs/issues/completed/ISSUE_665_parser-truncates-hyphenated-elements.md]: Fixed in f2ac91b. Updated the documentation regex snippet to show `[0-9]+[-+]` instead of `[0-9]+[-]`, matching the current implementation that allows both hyphen and plus as the first separator for number-starting identifiers.
